ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,450 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
 
-    if (DeviceKey != NULL)
-        DeviceKeyRealP = DeviceKey;
-    else
-        DeviceKeyRealP = &DeviceKeyReal;
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, DeviceString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
 
-    if (InstanceKey != NULL)
-        InstanceKeyRealP = InstanceKey;
+    Status = RtlAppendUnicodeToString(&SymbolicLink, L"#");
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    if (ReferenceString->Buffer != NULL)
+    {
+        Status = RtlAppendUnicodeStringToString(&SymbolicLink, ReferenceString);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+            return Status;
+        }
+    }
     else
-        InstanceKeyRealP = &InstanceKeyReal;
+    {
+        Status = RtlAppendUnicodeToString(&SymbolicLink, L"\\");
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+            return Status;
+        }
+    }
 
-    *GuidKeyRealP = NULL;
-    *DeviceKeyRealP = NULL;
-    *InstanceKeyRealP = NULL;
+    DPRINT("New symbolic link is '%wZ'\n", &SymbolicLink);
 
-    RtlInitUnicodeString(&BaseKeyU, BaseKeyString);
+    *SymbolicLinkName = SymbolicLink;
+    return STATUS_SUCCESS;
+}
 
-    /* Open the DeviceClasses key */
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &BaseKeyU,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               NULL,
-                               NULL);
-    Status = ZwOpenKey(&ClassesKey,
-                       DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                       &ObjectAttributes);
-    if (!NT_SUCCESS(Status))
+static
+NTSTATUS
+IopSeparateSymbolicLink(IN PUNICODE_STRING SymbolicLinkName,
+                        OUT OPTIONAL PUNICODE_STRING PrefixString,
+                        OUT OPTIONAL PUNICODE_STRING MungedString,
+                        OUT OPTIONAL PUNICODE_STRING GuidString,
+                        OUT OPTIONAL PUNICODE_STRING ReferenceString,
+                        OUT OPTIONAL PBOOLEAN ReferenceStringPresent,
+                        OUT OPTIONAL LPGUID InterfaceClassGuid)
+{
+    UNICODE_STRING KernelModePrefix = RTL_CONSTANT_STRING(L"\\??\\");
+    UNICODE_STRING UserModePrefix = RTL_CONSTANT_STRING(L"\\\\?\\");
```

Comment:
Where would you imagine this prefix would come from? No user-mode path should ever make it into these functions as far as I can tell.
(note, if you really think these functions accept this, this is something your kmtest can check)

Reply:
Hmm, as it appears, such symbolic link also can be retrieved from user-mode (e. g. by wdmaud.drv in case with audio), so user-mode prefix probably is valid too. At least my recent WinDBG analysis shows something interesting. When MS wdmaud.drv validates the device string (which appears to be a symbolic link), it passes that to its internal validation routine. The main thing I noticed, is the prefix: it looks as user-mode (obviously because it's called from user space):
![windbg-analysis](https://user-images.githubusercontent.com/26385117/141095248-c11c35a8-5698-4477-a4b3-f4de405dd9d3.png)
Although it might come from setupapi instead, so this proof is not enough I think. And I still need to check it via my kmtest.

Reply:
Yes @HBelusca. I already have it in my #3595 PR. Feel free to review it also. :slightly_smiling_face: